docs/01-app/03-api-reference/05-config/01-next-config-js/cacheHandlers.mdx

```diff
@@ -0,0 +1,351 @@
+---
+title: cacheHandlers
+description: Configure custom cache handlers for use cache directives in Next.js.
+related:
+  title: Related
+  description: View related API references.
+  links:
+    - app/api-reference/directives/use-cache
+    - app/api-reference/directives/use-cache-remote
+    - app/api-reference/directives/use-cache-private
+    - app/api-reference/config/next-config-js/cacheLife
+---
+
+The `cacheHandlers` configuration allows you to define custom cache storage implementations for [`'use cache'`](/docs/app/api-reference/directives/use-cache) and [`'use cache: remote'`](/docs/app/api-reference/directives/use-cache-remote). This enables you to store cached components and functions in external services or customize the caching behavior.
+
+> **Good to know**: The `cacheHandlers` (plural) configuration is specifically for `'use cache'` directives. This is different from `cacheHandler` (singular), which is used for ISR and App/Pages Router cache operations. Note that [`'use cache: private'`](/docs/app/api-reference/directives/use-cache-private) is not configurable through `cacheHandlers`.
+
+## When to use custom cache handlers
+
+The default in-memory cache is isolated to each Next.js instance. When self-hosting with:
+
+- **Multiple containers or instances** - Each instance has its own cache, causing misses when requests hit different servers
+- **Horizontal scaling** - Load balancers distribute requests across servers without shared cache
+- **Container orchestration** (Kubernetes, Docker Swarm, ECS) - Pod/container restarts lose the in-memory cache
+
+Custom cache handlers allow you to use shared storage (Redis, Memcached, DynamoDB, etc.) so cache is available across all instances and survives restarts.
+
+Custom handlers are also useful for:
+
+- Persistent cache across deployments
```

Comment:
Is this accurate?

Cache keys are currently namespaced with the build id:

<img width="670" height="86" alt="Image" src="https://github.com/user-attachments/assets/92c287a3-cd82-4db5-a1da-c719c896701e" />

Reply:
We'd love to be able to do this, don't get me wrong. I guess someone could parse the key and work around it, but keys are supposed to be opaque and understandably so.

It's a great to have feature, until it isn't (schema changes).

In an ideal world, instead of namespacing by build id another concept could be introduced. Maybe a "schema version". Cached data could then be safely shared across deployments.